pandas/core/algorithms.py

```diff
@@ -619,13 +619,19 @@ def factorize(values, sort=False, order=None, na_sentinel=-1, size_hint=None):
 
     if sort and len(uniques) > 0:
         from pandas.core.sorting import safe_sort
-        try:
-            order = uniques.argsort()
-            order2 = order.argsort()
-            labels = take_1d(order2, labels, fill_value=na_sentinel)
-            uniques = uniques.take(order)
-        except TypeError:
-            # Mixed types, where uniques.argsort fails.
+        if na_sentinel == -1:
+            # GH-25409 take_1d only works for na_sentinels of -1
+            try:
+                order = uniques.argsort()
+                order2 = order.argsort()
+                labels = take_1d(order2, labels, fill_value=na_sentinel)
+                uniques = uniques.take(order)
+            except TypeError:
```

Comment:
woa? this is way repetitive. why is what you have in the try not enough here?

Reply:
`take_1d` cannot handle custom na_sentinels. See the discussion in the issue.

Reply:
I read the issue. this is fix is way too hacky. A custom sentinel is not a blocking issue for 0.24.2

Reply:
Well, as I said above, I can also simply remove the if/else and try/except completely, by not using `take_1d` if that is preferred. A small performance gain is not worth introducing (now knowingly) a regression in behaviour.

But personally, I also have no problem with just merging this PR as is.

Reply:
this is just adding technical debt and am -1 in merging as is

Reply:
Can you then give an indication with what you would be OK, instead of just saying -1? 
Eg I already mentioned that I can remove the usage of take_1d all together (so how it was before 0.23)

Reply:
> Eg I already mentioned that I can remove the usage of take_1d all together (so how it was before 0.23)

this sounds fine,  though don't remember why this was changed.

Reply:
Some reasoning is explained here: https://github.com/pandas-dev/pandas/pull/19938#discussion_r173567099 (mainly performance I think)

Reply:
ok, then let's just defer this change to 0.25.0 and do this w/o the multiple repetitive try/excepts.

Reply:
If you don't like the extra if/else, but are OK with reverting to only `safe_sort` bu only for 0.25.0: would you be OK with the current PR for 0.24.2, if I directly do a follow-up PR to remove the if/else try/except on master / for 0.25.0 ?

Reply:
@jreback OK?

Reply:
This would be good for 0.24.2.

What's the duplicate code / tech debt here here? The call to `safe_sort`?